modules/images/webp-uploads/load.php

```diff
@@ -5,40 +5,294 @@
+			continue;
+		}
+
+		$current_size = $image_sizes[ $size_name ];
+		$sources      = array();
+		if ( array_key_exists( 'sources', $current_size ) && is_array( $current_size['sources'] ) ) {
+			$sources = $current_size['sources'];
+		}
+
+		// Try to find the mime type of the image size.
+		$current_mime = '';
+		if ( array_key_exists( 'mime-type', $current_size ) ) {
+			$current_mime = $current_size['mime-type'];
+		} elseif ( array_key_exists( 'file', $current_size ) ) {
+			$current_mime = wp_check_filetype( $current_size['file'] )['type'];
+		}
+
+		if ( empty( $current_mime ) ) {
+			$sizes[ $size_name ] = $current_size;
+			continue;
+		}
+
+		$sources[ $current_mime ] = array(
+			'file'     => array_key_exists( 'file', $current_size ) ? $current_size['file'] : '',
+			'filesize' => 0,
+		);
+
+		// Set the filesize from the current mime image.
+		$file_location = path_join( $dirname, $sources[ $current_mime ]['file'] );
+		if ( file_exists( $file_location ) ) {
+			$sources[ $current_mime ]['filesize'] = filesize( $file_location );
+		}
+
+		$formats = array_key_exists( $current_mime, $valid_mime_transforms )
+			? $valid_mime_transforms[ $current_mime ]
+			: array();
+
+		foreach ( $formats as $mime ) {
+			wp_schedule_single_event( time(), 'webp_uploads_create_image', array( $attachment_id, $size_name, $mime ) );
+		}
+
+		$current_size['sources'] = $sources;
+		$sizes[ $size_name ]     = $current_size;
+	}
+
+	$metadata['sizes'] = $sizes;
+
+	return $metadata;
+}
+
+add_filter( 'wp_generate_attachment_metadata', 'webp_uploads_create_sources_property', 10, 2 );
+
+/**
+ * Creates a new image based of the specified attachment with a defined mime type
+ * this image would be stored in the same place as the provided size name inside the
+ * metadata of the attachment.
+ *
+ * @since n.e.x.t
+ *
+ * @param int    $attachment_id The ID of the attachment we are going to use as a reference to create the image.
+ * @param string $size          The size name that would be used to create this image, out of the registered subsizes.
+ * @param string $mime          A mime type we are looking to use to create this image.
+ *
+ * @return bool|int|WP_Error
+ */
+function webp_uploads_generate_image_size( $attachment_id, $size, $mime ) {
+	$sizes    = wp_get_registered_image_subsizes();
+	$metadata = wp_get_attachment_metadata( $attachment_id );
+
+	if (
+		! isset( $metadata['sizes'][ $size ], $sizes[ $size ] )
+		|| ! is_array( $metadata['sizes'][ $size ] )
+		|| ! is_array( $sizes[ $size ] )
+	) {
+		return new WP_Error( 'image_mime_type_invalid_metadata', __( 'The image does not have a valid metadata.', 'performance-lab' ) );
+	}
+
+	// All subsizes are created out of the attached file.
+	$file = get_attached_file( $attachment_id );
+
+	// File does not exist.
+	if ( ! file_exists( $file ) ) {
+		return new WP_Error( 'image_file_size_not_found', __( 'The provided size does not have a valid image file.', 'performance-lab' ) );
+	}
```

Comment:
This is a bit unnecessary. How about we make the `$file` a required parameter of the function, since we have it available outside anyway? That's also similar to WP core's `_wp_make_subsizes` where the file is passed to as well.

Reply:
That could be a possibility, however by introducing a new parameter we are allowing external callers to define a different base image for the additional mime types, and in this particular case we want to make sure the additional mime types are created out of the attached file.

In addition to that the current API allows a simplified call from external callers to create new additional mime types without much complexity.

Examples:

- https://github.com/WordPress/performance/blob/0e94c15cc5d6164e14950c19847d370738422190/tests/modules/images/webp-uploads/webp-uploads-test.php#L229
- https://github.com/WordPress/performance/blob/0e94c15cc5d6164e14950c19847d370738422190/tests/modules/images/webp-uploads/webp-uploads-test.php#L246
- https://github.com/WordPress/performance/blob/0e94c15cc5d6164e14950c19847d370738422190/tests/modules/images/webp-uploads/webp-uploads-test.php#L261

Reply:
@mitogh That's a fair point. The decision depends a bit on whether we consider this function a "public" API function, or whether it's more for internal usage, like WP core's `_wp_make_subsizes()` - if it's the latter, my suggestion would make sense.

Given that this is currently more of a "public" API function, I agree we can leave this as is - the overhead of looking up the attached file should be minimal.